### ISSUE TITLE
various test failures on 32 bit x86 Alpine linux

### ISSUE THREAD

**Issue by ncopa** [cython/cython#3620]
cython 0.29.18 introduces various test failures on 32bit x86 on Alpine Linux:

```
FAIL: unpacked_struct_with_arrays (buffmt.__test__)                                                                     
Doctest: buffmt.__test__.unpacked_struct_with_arrays                                                                    
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                      
  File "/usr/lib/python3.8/doctest.py", line 2204, in runTest                                                           
    raise self.failureException(self.format_failure(new.getvalue()))                                                    
AssertionError: Failed doctest test for buffmt.__test__.unpacked_struct_with_arrays                                     
  File "/home/ncopa/aports/main/cython/src/cython-0.29.18/TEST_TMP/buffers/c/buffmt/buffmt.cpython-38-i386-linux-gnu.so"
, line unknown line number, in unpacked_struct_with_arrays                                                              
                                                                                                                        
----------------------------------------------------------------------                                                  
File "/home/ncopa/aports/main/cython/src/cython-0.29.18/TEST_TMP/buffers/c/buffmt/buffmt.cpython-38-i386-linux-gnu.so", 
line ?, in buffmt.__test__.unpacked_struct_with_arrays                                                                  
Failed example:                                                                                                         
    unpacked_struct_with_arrays("T{i:a:(8)f:b:f:c:Q:d:(5)i:e:i:f:i:g:xxxx(4)d:h:i:i:}")                                 
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest buffmt.__test__.unpacked_struct_with_arrays[0]>", line 1, in <module>
        unpacked_struct_with_arrays("T{i:a:(8)f:b:f:c:Q:d:(5)i:e:i:f:i:g:xxxx(4)d:h:i:i:}")
      File "tests/buffers/buffmt.pyx", line 448, in buffmt.unpacked_struct_with_arrays (buffmt.c:5361)
        cdef object[UnpackedStructWithArrays] buf = MockBuffer(
    ValueError: Buffer dtype mismatch; next field is at offset 80 but 76 expected


======================================================================
FAIL: unpacked_struct_with_arrays (buffmt)
Doctest: buffmt.unpacked_struct_with_arrays
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/doctest.py", line 2204, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for buffmt.unpacked_struct_with_arrays
  File "/home/ncopa/aports/main/cython/src/cython-0.29.18/TEST_TMP/buffers/c/buffmt/buffmt.cpython-38-i386-linux-gnu.so"
, line unknown line number, in unpacked_struct_with_arrays

----------------------------------------------------------------------
File "/home/ncopa/aports/main/cython/src/cython-0.29.18/TEST_TMP/buffers/c/buffmt/buffmt.cpython-38-i386-linux-gnu.so", 
line ?, in buffmt.unpacked_struct_with_arrays
Failed example:
    unpacked_struct_with_arrays("T{i:a:(8)f:b:f:c:Q:d:(5)i:e:i:f:i:g:xxxx(4)d:h:i:i:}")
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest buffmt.unpacked_struct_with_arrays[0]>", line 1, in <module>
        unpacked_struct_with_arrays("T{i:a:(8)f:b:f:c:Q:d:(5)i:e:i:f:i:g:xxxx(4)d:h:i:i:}")
      File "tests/buffers/buffmt.pyx", line 448, in buffmt.unpacked_struct_with_arrays (buffmt.c:5361)
        cdef object[UnpackedStructWithArrays] buf = MockBuffer(
    ValueError: Buffer dtype mismatch; next field is at offset 80 but 76 expected


======================================================================
FAIL: unpacked_struct_with_arrays (buffmt.__test__)
Doctest: buffmt.__test__.unpacked_struct_with_arrays
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/doctest.py", line 2204, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for buffmt.__test__.unpacked_struct_with_arrays
  File "/home/ncopa/aports/main/cython/src/cython-0.29.18/TEST_TMP/buffers/cpp/buffmt/buffmt.cpython-38-i386-linux-gnu.s
o", line unknown line number, in unpacked_struct_with_arrays 

----------------------------------------------------------------------
File "/home/ncopa/aports/main/cython/src/cython-0.29.18/TEST_TMP/buffers/cpp/buffmt/buffmt.cpython-38-i386-linux-gnu.so"
, line ?, in buffmt.__test__.unpacked_struct_with_arrays
Failed example:
    unpacked_struct_with_arrays("T{i:a:(8)f:b:f:c:Q:d:(5)i:e:i:f:i:g:xxxx(4)d:h:i:i:}")
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest buffmt.__test__.unpacked_struct_with_arrays[0]>", line 1, in <module>
        unpacked_struct_with_arrays("T{i:a:(8)f:b:f:c:Q:d:(5)i:e:i:f:i:g:xxxx(4)d:h:i:i:}")
      File "tests/buffers/buffmt.pyx", line 448, in buffmt.unpacked_struct_with_arrays (buffmt.cpp:5375)
        cdef object[UnpackedStructWithArrays] buf = MockBuffer(
    ValueError: Buffer dtype mismatch; next field is at offset 80 but 76 expected


======================================================================
FAIL: unpacked_struct_with_arrays (buffmt)
Doctest: buffmt.unpacked_struct_with_arrays
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/doctest.py", line 2204, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for buffmt.unpacked_struct_with_arrays
  File "/home/ncopa/aports/main/cython/src/cython-0.29.18/TEST_TMP/buffers/cpp/buffmt/buffmt.cpython-38-i386-linux-gnu.s
o", line unknown line number, in unpacked_struct_with_arrays 

----------------------------------------------------------------------
File "/home/ncopa/aports/main/cython/src/cython-0.29.18/TEST_TMP/buffers/cpp/buffmt/buffmt.cpython-38-i386-linux-gnu.so"
, line ?, in buffmt.unpacked_struct_with_arrays
Failed example:
    unpacked_struct_with_arrays("T{i:a:(8)f:b:f:c:Q:d:(5)i:e:i:f:i:g:xxxx(4)d:h:i:i:}")
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest buffmt.unpacked_struct_with_arrays[0]>", line 1, in <module>
        unpacked_struct_with_arrays("T{i:a:(8)f:b:f:c:Q:d:(5)i:e:i:f:i:g:xxxx(4)d:h:i:i:}")
      File "tests/buffers/buffmt.pyx", line 448, in buffmt.unpacked_struct_with_arrays (buffmt.cpp:5375)
        cdef object[UnpackedStructWithArrays] buf = MockBuffer(
    ValueError: Buffer dtype mismatch; next field is at offset 80 but 76 expected


----------------------------------------------------------------------
Ran 11873 tests in 6798.719s

FAILED (failures=4, skipped=34)
```

this does not happen on x86_64.

**Comment by scoder** [comment]
Thanks for the report. This is due to #3562 and AFAICT, should only be a non-portable test assumption and not wrong code being generated.

@smutch I think we need to revise this patch, again.

**Comment by smutch** [comment]
Sorry this has crept up. I've just issued a PR to try and address the failures.

Unfortunately I don't have access to a 32bit OS to test this on.  I attempted to use an i386 (32 bit) Alpine linux docker image but was unable to get things running, at least with miniconda.  @ncopa do you have a Dockerfile that you would be willing to share which would allow me to test the fix?

**Comment by ncopa** [comment]
Paste this into your console:

```sh
cat >Dockerfile <<EOF
FROM i386/alpine:edge

ENV pkgver=0.29.18

RUN apk upgrade -U -a \
	&& apk add py3-setuptools python3-dev build-base

RUN mkdir /src \
	&& cd /src \
	&& wget https://github.com/cython/cython/archive/\$pkgver.tar.gz \
	&& tar -zxf \$pkgver.tar.gz
RUN cd /src/cython-\$pkgver \
	&& linux32 python3 setup.py build
EOF
docker build -t cython3620 .
docker run --rm -it cython3620 sh -c "cd /src/cython-0.29.18 && PYTHON=python3 linux32 make test"
```

**Comment by ncopa** [comment]
Is there some way i can re-run only the failing tests?
I could test https://github.com/cython/cython/pull/3621

**Comment by ncopa** [comment]
> Is there some way i can re-run only the failing tests?
> I could test #3621

I think I figured it out: `make test TESTOPTS=buffmt PYTHON=python3`

**Comment by scoder** [comment]
Resolved in #3621.
